pkg/tools/etcd_helper.go

```diff
@@ -116,19 +127,70 @@ func (h *EtcdHelper) decodeNodeList(nodes []*etcd.Node, slicePtr interface{}) er
 			}
 			continue
 		}
-		obj := reflect.New(v.Type().Elem())
-		if err := h.Codec.DecodeInto([]byte(node.Value), obj.Interface().(runtime.Object)); err != nil {
-			return err
-		}
-		if h.Versioner != nil {
-			// being unable to set the version does not prevent the object from being extracted
-			_ = h.Versioner.UpdateObject(obj.Interface().(runtime.Object), node)
+		if obj, found := h.getFromCache(node.ModifiedIndex); found {
+			v.Set(reflect.Append(v, reflect.ValueOf(obj)))
+		} else {
+			obj := reflect.New(v.Type().Elem())
+			if err := h.Codec.DecodeInto([]byte(node.Value), obj.Interface().(runtime.Object)); err != nil {
+				return err
+			}
+			if h.Versioner != nil {
+				// being unable to set the version does not prevent the object from being extracted
+				_ = h.Versioner.UpdateObject(obj.Interface().(runtime.Object), node)
+			}
+			v.Set(reflect.Append(v, obj.Elem()))
+			if node.ModifiedIndex != 0 {
+				h.addToCache(node.ModifiedIndex, obj.Interface())
+			}
 		}
-		v.Set(reflect.Append(v, obj.Elem()))
 	}
 	return nil
 }
 
+// etcdCache defines interface used for caching objects stored in etcd. Objects are keyed by
+// their Node.ModifiedIndex, which is unique across all types.
+// All implementations must be thread-safe.
+type etcdCache interface {
+	getFromCache(index uint64) (interface{}, bool)
+	addToCache(index uint64, obj interface{})
+}
+
+const maxEtcdCacheEntries int = 100000
+
+func (h *EtcdHelper) getFromCache(index uint64) (interface{}, bool) {
+	var obj interface{}
```

Comment:
> On Apr 28, 2015, at 7:21 AM, Filip Grzadkowski <notifications@github.com> wrote:
> 
> In pkg/tools/etcd_helper.go:
> 
> >  	}
> >  	return nil
> >  }
> >  
> > +// etcdCache defines interface used for caching objects stored in etcd. Objects are keyed by
> > +// their Node.ModifiedIndex, which is unique across all types.
> > +// All implementations must be thread-safe.
> > +type etcdCache interface {
> > +	getFromCache(index uint64) (interface{}, bool)
> > +	addToCache(index uint64, obj interface{})
> > +}
> > +
> > +const maxEtcdCacheEntries int = 100000
> > +
> > +func (h *EtcdHelper) getFromCache(index uint64) (interface{}, bool) {
> > +	var obj interface{}
> @smarterclayton I checked memory and (surprisingly) memory footprint is much lower with cache. Maximum reserved memory during performance tests (100 nodes, 3000 pods) is:
> 
> 2.9 GB without cache
> 1.8 GB with cache It seems that parsing JSON uses a lot of memory. If it's done in multliple goroutines concurrently it can affect memory usage pretty significantly
> @lavalamp I think that even with fast conversions we will still benefit from cache due to:
> 
> reduce memory footprint
> unmarshalling JSON can actually use a lot of CPU (I've seen profiles with >20%)
I'd like I see a quick investigation of ugorji after the conversion work is in place - if unmarshalling JSON is making a significant memory impact and ugorji is cheap to setup (since I believe it can fall back to the default json serializer implementation), it may be a very quick win for us.
> —
> Reply to this email directly or view it on GitHub.
>